services/storage/tests/unit/test_handlers_files.py

```diff
@@ -962,3 +965,351 @@ async def test_copy_as_soft_link(
+                    f"--> done waiting, data is completely uploaded [{attempt.retry_state.retry_object.statistics}]"
+                )
+
+        return directory_file_upload
+
+    return _directory_creator
+
+
+@pytest.fixture
+async def populate_directory(
+    create_file_of_size: Callable[[ByteSize, str | None], Path],
+    storage_s3_client: StorageS3Client,
+    storage_s3_bucket: S3BucketName,
+    project_id: ProjectID,
+    node_id: NodeID,
+) -> Callable[..., Awaitable[None]]:
+    async def _create_content(
+        file_size_in_dir: ByteSize,
+        dir_name: str,
+        subdir_count: int = 4,
+        file_count: int = 5,
+    ) -> None:
+        file = create_file_of_size(file_size_in_dir, "some_file")
+
+        async def _create_file(s: int, f: int):
+            file_name = f"{dir_name}/sub-dir-{s}/file-{f}"
+            clean_path = Path(f"{project_id}/{node_id}/{file_name}")
+            await storage_s3_client.upload_file(
+                storage_s3_bucket, file, SimcoreS3FileID(f"{clean_path}"), None
+            )
+
+        tasks: deque = deque()
+        for s in range(subdir_count):
+            for f in range(file_count):
+                tasks.append(_create_file(s, f))
+
+        await asyncio.gather(*tasks)
+
+        file.unlink()
+
+    return _create_content
+
+
+@pytest.fixture
+async def delete_directory(
+    client: TestClient, user_id: UserID, location_id: LocationID
+) -> Callable[..., Awaitable[None]]:
+    async def _dir_remover(directory_file_upload: FileUploadSchema) -> None:
+        directory_file_id = directory_file_upload.urls[0].path.strip("/")
+        delete_url = (
+            client.app.router["delete_file"]
+            .url_for(
+                location_id=f"{location_id}",
+                file_id=urllib.parse.quote(directory_file_id, safe=""),
+            )
+            .with_query(user_id=user_id)
+        )
+        response = await client.delete(f"{delete_url}")
+        await assert_status(response, web.HTTPNoContent)
+
+    return _dir_remover
+
+
+@pytest.fixture
+async def directory_with_files(
+    create_empty_directory: Callable[..., Awaitable[FileUploadSchema]],
+    populate_directory: Callable[..., Awaitable[None]],
+    delete_directory: Callable[..., Awaitable[None]],
+) -> Callable[..., AsyncContextManager[FileUploadSchema]]:
+    @asynccontextmanager
+    async def _context_manager(
+        dir_name: str, file_size_in_dir: ByteSize, subdir_count: int, file_count: int
+    ) -> AsyncIterator[FileUploadSchema]:
+        directory_file_upload: FileUploadSchema = await create_empty_directory(
+            dir_name=dir_name
+        )
+
+        await populate_directory(
+            file_size_in_dir=file_size_in_dir,
+            dir_name=dir_name,
+            subdir_count=subdir_count,
+            file_count=file_count,
+        )
+
+        yield directory_file_upload
+
+        await delete_directory(directory_file_upload=directory_file_upload)
```

Comment:
here you do not check if the folder is really empty. my feeling is that you have still some files around. Can you please add an assertion for this?